### PR TITLE
Add samples for building Java Modules

### PR DESCRIPTION
**Pull request by jjohannes**

### CODE REVIEW

**Comment by jjohannes** [comment]
@bot-gradle test QF

File: gradle/wrapper/gradle-wrapper.properties [hunk 1]
```diff
@@ -1,5 +1,5 @@
 distributionBase=GRADLE_USER_HOME
 distributionPath=wrapper/dists
-distributionUrl=https\://services.gradle.org/distributions-snapshots/gradle-6.4-20200319134018+0000-bin.zip
+distributionUrl=https\://services.gradle.org/distributions-snapshots/gradle-6.4-20200401220055+0000-bin.zip
```
Comment (bmuskalla) [inline gradle/wrapper/gradle-wrapper.properties:3]: Is this change intentional?

Comment (bmuskalla) [inline gradle/wrapper/gradle-wrapper.properties:3]: Is this change intentional?

File: subprojects/docs/docs.gradle [hunk 2]
```diff
@@ -147,6 +147,23 @@ samples {
...
+        }
+        javaModuleInfoForApplication {
+            target = "application"
+        }
+        javaJunit5ModuleInfoForUtilitiesLibrary {
+            target = "utilities"
+        }
+        javaJunit5ModuleInfoForApplication {
+            target = "application"
+        }
```
Comment (bmuskalla) [inline subprojects/docs/docs.gradle:164]: I see that this best matches the structure but it just hurts to see that we need more boilerplate code for those files than the files have content. Do you think we should merge them into one template?

File: subprojects/docs/src/samples/java/modules-multi-project-with-integration-tests/README.adoc [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+```
+open module org.gradle.sample.integtest.utilities {
+    requires org.gradle.sample.utilities;
+    requires org.junit.jupiter.api;
+}
+```
+
+NOTE: This sample does **not** work in Eclipse when imported using https://projects.eclipse.org/projects/tools.buildship[Eclipse Buildship].
+This is due to a limitation in Eclipse that does not allow more than one module in one project.
```
Comment (bmuskalla) [inline subprojects/docs/src/samples/java/modules-multi-project-with-integration-tests/README.adoc:25]: For what its worth, it may be nice to link the corresponding eclipse bug.

File: subprojects/docs/src/samples/java/modules-multi-project-with-integration-tests/kotlin/settings.gradle.kts [hunk 4]
```diff
@@ -0,0 +1,2 @@
+rootProject.name = "components-with-junit-5-tests"
```
Comment (bmuskalla) [inline subprojects/docs/src/samples/java/modules-multi-project-with-integration-tests/kotlin/settings.gradle.kts:1]: wrong name

File: subprojects/docs/src/samples/java/modules-multi-project/kotlin/application/build.gradle.kts [hunk 5]
```diff
@@ -0,0 +1,14 @@
+plugins {
+    application
+}
+
+dependencies {
+    implementation(project(":utilities"))
+    testImplementation("org.junit.jupiter:junit-jupiter-api:5.6.0")
```
Comment (bmuskalla) [inline subprojects/docs/src/samples/java/modules-multi-project/kotlin/application/build.gradle.kts:7]: :o: all other samples use 5.6.1

File: subprojects/docs/src/samples/java/modules-multi-project/kotlin/settings.gradle.kts [hunk 6]
```diff
@@ -0,0 +1,2 @@
+rootProject.name = "components-with-junit-5-tests"
```
Comment (bmuskalla) [inline subprojects/docs/src/samples/java/modules-multi-project/kotlin/settings.gradle.kts:1]: :x: wrong name

File: gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (jjohannes) [reply gradle/wrapper/gradle-wrapper.properties:3]: Yes. There is an issue at the moments that the samples are tested against this ☝️ Gradle versions and not the current version under development. The previous wrapper version here did not contain JPMS support.

File: subprojects/docs/src/samples/java/modules-multi-project-with-integration-tests/README.adoc (hunk 3 above)
Comment (jjohannes) [reply subprojects/docs/src/samples/java/modules-multi-project-with-integration-tests/README.adoc:25]: Good point. Will do.

File: gradle/wrapper/gradle-wrapper.properties (hunk 1 above)
Comment (jjohannes) [reply gradle/wrapper/gradle-wrapper.properties:3]: Yes. The samples are at the moment tested against the currently running Gradle version. So we need a wrapper that includes the JPMS support.

File: subprojects/docs/docs.gradle (hunk 2 above)
Comment (jjohannes) [reply subprojects/docs/docs.gradle:164]: I see your point. I think I like to keep following the structure for now and revisit this when we added more JVM samples. I think at one point we should at least organize the `templates` folder with subfolders. And maybe there is also a way to improve this 'sample definition' DSL so that we can keep the structure, but have less boilerplate to write. (E.g. If we have a common structure, we can maybe just read the folder without having to explicitly define templates)
